python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,691 @@
+            indices = np.where(sumImage.image.array > 0)
+        else:
+            if mode == 'AND':  # must appear in all
+                threshold = 1
+            elif mode == 'FRACTION':
+                threshold = self.config.combinationFraction
+            else:
+                raise RuntimeError(f"Got unsupported combinationMode {mode}")
+            indices = np.where(sumImage.image.array >= threshold)
+
+        BADBIT = sumImage.mask.getPlaneBitMask('BAD')
+        sumImage.mask.array[indices] |= BADBIT
+
+        self.log.info("Post-merge %s pixels marked as defects" % len(indices[0]))
+
+        if self.config.edgesAsDefects:
+            self.log.info("Masking edge pixels as defects in addition to previously identified defects")
+            self._setEdgeBits(sumImage, 'BAD')
+
+        defects = measAlg.Defects.fromMask(sumImage, 'BAD')
+        return defects
+
+    @staticmethod
+    def _getNumGoodPixels(maskedIm, badMaskString="NO_DATA"):
+        """Return the number of non-bad pixels in the image."""
+        nPixels = maskedIm.mask.array.size
+        nBad = countMaskedPixels(maskedIm, badMaskString)
+        return nPixels - nBad
+
+    def findHotAndColdPixels(self, exp, imageType, setMask=False):
+        """Find hot and cold pixels in an image.
+
+        Using config-defined thresholds on a per-amp basis, mask pixels
+        that are nSigma above threshold in dark frames (hot pixels),
+        or nSigma away from the clipped mean in flats (hot & cold pixels).
+
+        Parameters
+        ----------
+        exp : `lsst.afw.image.exposure.Exposure`
+            The exposure in which to find defects.
+        imageType : `str`
+            The image type, either 'dark' or 'flat'.
+        setMask : `bool`
+            If true, update exp with hot and cold pixels.
+            hot: DETECTED
+            cold: DETECTED_NEGATIVE
+
+        Returns
+        -------
+        defects : `lsst.meas.algorithms.Defect`
+            The defects found in the image.
+        """
+        assert imageType in ['flat', 'dark']
+
+        self._setEdgeBits(exp)
+        maskedIm = exp.maskedImage
+
+        # the detection polarity for afwDetection, True for positive,
+        # False for negative, and therefore True for darks as they only have
+        # bright pixels, and both for flats, as they have bright and dark pix
+        polarities = {'dark': [True], 'flat': [True, False]}[imageType]
+
+        footprintList = []
+
+        for amp in exp.getDetector():
+            ampImg = maskedIm[amp.getBBox()].clone()
+
+            # crop ampImage depending on where the amp lies in the image
+            if self.config.nPixBorderLeftRight:
+                if ampImg.getX0() == 0:
+                    ampImg = ampImg[self.config.nPixBorderLeftRight:, :, afwImage.LOCAL]
+                else:
+                    ampImg = ampImg[:-self.config.nPixBorderLeftRight, :, afwImage.LOCAL]
+            if self.config.nPixBorderUpDown:
```

Comment:
There's almost certainly a method on amp to get the BBox (and you should use it).  Which pixels are you looking for here?

Reply:
Many points here:
* This isn't working on the amps, this is working on the whole detector
* There are indeed methods to get the bbox, and yes, you could then dilate/contract it, and you could make copies, and do gross stuff ORing them or whatever, but it's going to be horrible
* Even then I don't see how you'd easily get a rectangle on each edge easily
* And if you somehow did, it would be really muddy to them break the symmetry to support different sized borders at the top/bottom vs the left/right

So yes, there probably _is_ a way to do it with bboxes, but I think it's going to be much worse than what is currently here.

Reply:
Basically, applying asymmetric amp-wise borders is not something I know how to do elegantly, and doesn't come up that often I don't think. I do agree that it should be supported by lower level code, but moving that stuff there is well beyond the scope of this ticket.